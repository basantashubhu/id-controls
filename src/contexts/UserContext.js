import {createContext, useContext, useEffect, useState} from "react";
import {auth} from '../firebase'

export const UserContext = createContext(null);

export function useAuth() {
    return useContext(UserContext)
}

export function AuthProvider({children}) {
    const [currentUser, setCurrentUser] = useState()
    const [loading, setLoading] = useState(false)
    const signUp = (email, password) => {
        return auth.createUserWithEmailAndPassword(email, password);
    }
    const logout = () => {
        return auth.signOut()
    }
    const login = (email, password) => {
        return auth.signInWithEmailAndPassword(email, password)
    }
    const resetPassword = (email) => {
        return auth.sendPasswordResetEmail(email)
    }
    useEffect(() => {
        return auth.onAuthStateChanged(user => {
            setCurrentUser(user);
            setLoading(false);
        })
    }, []);
    const value = {
        currentUser,
        signUp,
        logout,
        login,
        resetPassword
    };
    return <UserContext.Provider value={value}>
        {!loading && children}
    </UserContext.Provider>
}