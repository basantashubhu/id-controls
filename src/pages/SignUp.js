import {useState} from "react";
import {useAuth} from "../contexts/UserContext";
import {Link, useHistory} from "react-router-dom";

const SignUp = () => {
    const history = useHistory()
    const {signUp, addProfile} = useAuth()
    const [credential, setCredential] = useState({email: '', password: '', cpassword: '', first_name : '', last_name : ''})
    const [error, setError] = useState('')
    const [loading, setLoading] = useState(false)
    const handleChange = (e) => {
        setCredential({...credential, [e.target.name]: e.target.value})
    }
    const handleSignUp = async (e) => {
        e.preventDefault();
        if (credential.password !== credential.cpassword) {
            return setError('Password do not match');
        }
        try {
            setError('');
            setLoading(true);
            addProfile({
                first_name : credential.first_name,
                last_name : credential.last_name
            })
            await signUp(credential.email, credential.password);
            history.push('/')
        } catch (e) {
            setError('Failed to create an account');
        } finally {
            setLoading(false)
        }
    }
    return <>
        <div className="container">
            <div className="d-flex justify-content-center">
                <div className="w-100" style={{maxWidth: '400px'}}>
                    <div className="card">
                        <div className="card-body">
                            <h2 className="text-center mb-4">Sign Up</h2>
                            {error && <div className="alert alert-danger">{error}</div>}
                            <form onSubmit={handleSignUp}>
                                <div className="form-group">
                                    <label>First Name</label>
                                    <input type="text" name="first_name" className="form-control" value={credential.first_name}
                                           onChange={handleChange} required/>
                                </div>
                                <div className="form-group">
                                    <label>Last Name</label>
                                    <input type="text" name="last_name" className="form-control" value={credential.last_name}
                                           onChange={handleChange} required/>
                                </div>
                                <div className="form-group">
                                    <label>Email</label>
                                    <input type="email" name="email" className="form-control" value={credential.email}
                                           onChange={handleChange} required/>
                                </div>
                                <div className="form-group">
                                    <label>Password</label>
                                    <input type="password" name="password" className="form-control"
                                           value={credential.password} onChange={handleChange} required/>
                                </div>
                                <div className="form-group">
                                    <label>Password Confirmation</label>
                                    <input type="password" name="cpassword" className="form-control"
                                           value={credential.cpassword} onChange={handleChange} required/>
                                </div>
                                <div className="form-group">
                                    <button type="submit" className="btn btn-primary w-100" disabled={loading}>Sign Up</button>
                                </div>
                            </form>
                        </div>
                    </div>
                    <p className="text-center mt-3">Already have an account? <Link to={'/login'}>Login</Link></p>
                </div>
            </div>
        </div>
    </>
}

export default SignUp