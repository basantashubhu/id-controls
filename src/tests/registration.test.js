import React from 'react'
import {shallow, configure} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import SignUp from '../pages/SignUp'
import * as UserContext from '../contexts/UserContext'
configure({adapter :  new Adapter()})
describe('Test case for registration', () => {
    let wrapper, useAuth;
    beforeAll(() => {
        useAuth = jest.spyOn(UserContext, 'useAuth')
    });
    beforeEach(() => {
        useAuth.mockClear();
        useAuth.mockReturnValue(() => ({
            signUp : () => 1
        }))
    })
    it('first name check', async () => {
        wrapper = shallow(<SignUp/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('[name="first_name"]')).toHaveLength(1);
    })
    it('last name check', () => {
        wrapper = shallow(<SignUp/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('[name="last_name"]')).toHaveLength(1);
    })
    it('email check', () => {
        wrapper = shallow(<SignUp/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('[name="email"]')).toHaveLength(1);
    })
    it('password check', () => {
        wrapper = shallow(<SignUp/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('[name="password"]')).toHaveLength(1);
    })
    it('password confirmation check', () => {
        wrapper = shallow(<SignUp/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('[name="cpassword"]')).toHaveLength(1);
    })
})
