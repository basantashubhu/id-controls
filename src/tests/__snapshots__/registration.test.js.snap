// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test case for registration email check 1`] = `ShallowWrapper {}`;

exports[`Test case for registration first name check 1`] = `ShallowWrapper {}`;

exports[`Test case for registration last name check 1`] = `ShallowWrapper {}`;

exports[`Test case for registration password check 1`] = `ShallowWrapper {}`;

exports[`Test case for registration password confirmation check 1`] = `ShallowWrapper {}`;
